File: archiso/airootfs/usr/share/churros/popups/battery/widgets/battery.py
from pathlib import Path
import sys

ROOT = Path(__file__).resolve().parents[2]
PREFERENCES = Path(__file__).resolve().parents[4] / "preferences"

sys.path.insert(0, str(ROOT))
sys.path.insert(0, str(PREFERENCES))

import gi

gi.require_version("Gtk", "4.0")

from gi.repository import Gtk, GLib

from i18n import _

from services.battery import BatteryService


_STATE_LABELS = {
    "charging": "Charging",
    "discharging": "Discharging",
    "fully-charged": "Full",
    "pending-charge": "Pending charge",
    "pending-discharge": "Pending discharge",
    "empty": "Empty",
    "unknown": "Unknown",
}


class BatteryWidget(Gtk.Box):

    def __init__(self):

        super().__init__(
            orientation=Gtk.Orientation.VERTICAL,
            spacing=12
        )

        self.add_css_class("battery-widget")

        self.percentage = Gtk.Label()
        self.percentage.add_css_class("battery-percentage")

        self.status = Gtk.Label()
        self.status.add_css_class("battery-status")

        self.remaining = Gtk.Label()
        self.remaining.add_css_class("battery-remaining")

        self.append(self.percentage)
        self.append(self.status)
        self.append(self.remaining)

        self.update()

        GLib.timeout_add_seconds(
            5,
            self.update
        )

    def update(self):

        try:

            battery = BatteryService.get()

        except Exception:

            battery = {"available": False}

        if not battery["available"]:

            self.percentage.set_label(_("No battery detected"))
            self.status.set_visible(False)
            self.remaining.set_visible(False)

            return True

        self.status.set_visible(True)
        self.remaining.set_visible(True)

        self.percentage.set_label(
            f"{battery['icon']} {battery['percentage']}%"
        )

        state_key = battery["state"]
        state_label = _STATE_LABELS.get(state_key, state_key.title())
        self.status.set_label(_(state_label))

        if battery["state"] == "charging" and battery["time_to_full"]:

            self.remaining.set_label(
                f"{battery['time_to_full']} {_('until full')}"
            )

        elif battery["state"] == "discharging" and battery["time_to_empty"]:

            self.remaining.set_label(
                f"{battery['time_to_empty']} {_('until empty')}"
            )

        else:

            self.remaining.set_label("")

        return True

File: rust/popups/src/battery.rs
// ==========================================
// battery.rs — popup de batería (port de battery/window.py + widgets)
// ==========================================

use churros_services::battery;
use gtk::prelude::*;

use crate::popup::PopupWindow;

pub fn build(app: &gtk::Application) -> PopupWindow {
    let w = PopupWindow::new(app, "Battery", "󰁹", "battery.css");
    w.add(&battery_widget());
    w
}

fn state_label(state: &str) -> String {
    match state {
        "charging" => "Charging".to_string(),
        "discharging" => "Discharging".to_string(),
        "fully-charged" => "Full".to_string(),
        "pending-charge" => "Pending charge".to_string(),
        "pending-discharge" => "Pending discharge".to_string(),
        "empty" => "Empty".to_string(),
        "unknown" => "Unknown".to_string(),
        other => title_case(other),
    }
}

/// Equivalente a str.title() de Python.
fn title_case(s: &str) -> String {
    let mut out = String::with_capacity(s.len());
    let mut capitalize = true;
    for ch in s.chars() {
        if ch.is_alphanumeric() {
            if capitalize {
                out.extend(ch.to_uppercase());
                capitalize = false;
            } else {
                out.push(ch);
            }
        } else {
            out.push(ch);
            capitalize = true;
        }
    }
    out
}

/// Indicador de batería: icono + porcentaje, estado, tiempo y barra visual
/// (port de battery/widgets/battery.py + barra de progreso).
fn battery_widget() -> gtk::Box {
    let vbox = gtk::Box::new(gtk::Orientation::Vertical, 12);
    vbox.add_css_class("battery-widget");

    let row = gtk::Box::new(gtk::Orientation::Horizontal, 12);

    let icon = gtk::Label::new(None);
    icon.add_css_class("battery-icon");

    let percentage = gtk::Label::new(None);
    percentage.add_css_class("battery-percentage");

    row.append(&icon);
    row.append(&percentage);

    let status = gtk::Label::new(None);
    status.add_css_class("battery-status");
    status.set_halign(gtk::Align::Start);

    let bar = gtk::LevelBar::builder()
        .min_value(0.0)
        .max_value(100.0)
        .value(0.0)
        .build();
    bar.add_css_class("battery-bar");

    let remaining = gtk::Label::new(None);
    remaining.add_css_class("battery-remaining");
    remaining.set_halign(gtk::Align::Start);

    vbox.append(&row);
    vbox.append(&status);
    vbox.append(&bar);
    vbox.append(&remaining);

    update(&icon, &percentage, &status, &bar, &remaining);

    glib::timeout_add_seconds_local(
        5,
        glib::clone!(
            #[strong] icon,
            #[strong] percentage,
            #[strong] status,
            #[strong] bar,
            #[strong] remaining,
            move || {
                update(&icon, &percentage, &status, &bar, &remaining);
                glib::ControlFlow::Continue
            }
        ),
    );

    vbox
}

fn update(
    icon: &gtk::Label,
    pct: &gtk::Label,
    status: &gtk::Label,
    bar: &gtk::LevelBar,
    remaining: &gtk::Label,
) {
    let data = battery::get();

    if !data.available {
        icon.set_label("");
        pct.set_label("No battery detected");
        status.set_visible(false);
        bar.set_visible(false);
        remaining.set_visible(false);
        return;
    }

    status.set_visible(true);
    bar.set_visible(true);
    remaining.set_visible(true);

    icon.set_label(&data.icon);
    pct.set_label(&format!("{}%", data.percentage));
    bar.set_value(data.percentage as f64);

    let charging = data.state == "charging";
    if charging {
        status.add_css_class("charging");
        status.remove_css_class("low");
    } else if data.percentage < 20 {
        status.add_css_class("low");
        status.remove_css_class("charging");
    } else {
        status.remove_css_class("charging");
        status.remove_css_class("low");
    }
    status.set_label(&state_label(&data.state));

    if data.state == "charging" && !data.time_to_full.is_empty() {
        remaining.set_label(&format!("{} until full", data.time_to_full));
    } else if data.state == "discharging" && !data.time_to_empty.is_empty() {
        remaining.set_label(&format!("{} until empty", data.time_to_empty));
    } else {
        remaining.set_label("");
    }
}